### ISSUE TITLE
Index parsing errors are swallowed

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/cargo#14894]
### Problem

When using a private registry, I recently ran into a case where a crate version that _is_ in the index (confirmed by cURLing the registry directly) could not be resolved by Cargo, giving me an error like
```
error: no matching package named `something` found
location searched: `private-registry` index
```

I quickly suspected that the index entry was invalid somehow, but Cargo gave no errors or warnings, even with `CARGO_LOG=trace`. Furthermore, by tweaking `http_remote.rs`, I found that its query to the registry (with an `if-none-match` header) returned `NotModified`, indicating Cargo _did_ have the etag of the latest index file, which did include the crate/version in question. By removing the etag header, I also confirmed that the contents Cargo got from its request contained the version in question.

Long story short, I discovered that Cargo ends up swallowing errors related to invalid index entries. The first time Cargo downloads an index file with an invalid entry, it hits

https://github.com/rust-lang/cargo/blob/05f54fdc34310f458033af8a63ce1d699fae8bf6/src/cargo/core/summary.rs#L345-L348

However, that gets turned into a `tracing::info` in

https://github.com/rust-lang/cargo/blob/05f54fdc34310f458033af8a63ce1d699fae8bf6/src/cargo/sources/registry/index/mod.rs#L604-L619

The user is likely to miss this since Cargo tracing logs are off by default, so they will just observe that Cargo claims the version doesn't exist.

Unfortunately, the issue doesn't stop there. If the user _then_ tries to run with `CARGO_LOG=trace`, they will then see _no_ errors or warnings from Cargo. In fact, the logs will make it seem like the invalid entry doesn't exist at all. There will be no trace of it anywhere. This is because when Cargo parses an index entry, it caches the parsed lines so that it doesn't have to parse them again

https://github.com/rust-lang/cargo/blob/05f54fdc34310f458033af8a63ce1d699fae8bf6/src/cargo/sources/registry/index/mod.rs#L622

+

https://github.com/rust-lang/cargo/blob/05f54fdc34310f458033af8a63ce1d699fae8bf6/src/cargo/sources/registry/index/mod.rs#L627-L631

However, when a line can't be parsed, it doesn't cache that line (note the `continue` where the error gets turned into `tracing::info!`). It _also_ doesn't leave that line as "unparsed" somewhere. So on subsequent invocations, Cargo grabs the version list from the cache manager to avoid parsing lines again, doesn't observe the (invalid) index entry for the version in question so it doesn't warn, then downloads the file again because it doesn't find the requested version, that download yields and empty response because the etag _does_ match the latest index file, and so Cargo gives up.

So, two requests:

1. Cargo should be louder when it finds an invalid index entry.
2. Cargo should repeat the warning about an invalid index entry even if it's in an up-to-date cached version of the index file.

### Steps

1. Inject a new version with an invalid index entry into a registry.
2. Take a dependency on said version.
3. Run `cargo build`.
4. Observe that Cargo only says "version not found" with no further warnings.
5. Re-run `cargo build` with `CARGO_LOG=trace`.
6. Observe that there is no information in the logs about the invalid index entry which caused the version not to be found.

### Notes

In my case, the index entry parsing error I was eventually able to extract by disabling the `if-none-match` logic and then running with `CARGO_LOG=trace` was:

```
failed to parse "so/me/something" registry package: optional dependency `uom_0_35` is not included in any feature
Make sure that `dep:uom_0_35` is included in one of features in the [features] table.
```

I _suspect_, but am not quite sure, that this may be an erroneous index entry manipulation by the private registry provider we're using. Although it could also be that `cargo publish` allowed publishing something with a nonsensical optional dependency in the first place?

### Version

```text
cargo 1.85.0 (05f54fdc3 2024-12-03)
```

**Comment by epage** [comment]
This is heavily related to #10623.

Ideally we track the error state through Cargo.  We've added [`IndexSummary`](https://doc.rust-lang.org/nightly/nightly-rustc/cargo/sources/registry/enum.IndexSummary.html) to better track state.  That now tracks newer versions which can help with #10623 (though we should also track the MSRV if present) though we still need to pipe this through up to the user.  Also, unsure if that is only on successful parses or also on failed parses which we should also support.

What is distinct about this Issue compared to #10623 is adding an error variant for when the parse fails for a reason other than a newer index version.

Granted, this will only help if we can extract the name and version.  If its too corrupted to even do that, we'll need to consider other means of reporting.

**Comment by ehuss** [comment]
Thanks for the report! The failures are silenced somewhat on purpose, since this is intended to handle the scenario where an older version doesn't understand the format of newer index entries. We wanted to avoid the situation where older versions would spam warnings on the console whenever they are used. I think this is somewhat related or a duplicate of #7929 where we could have a controlled method of presenting this information. Or maybe similarly to #4309, where there could be a separate command to display that.

**Comment by epage** [comment]
I think there are two separate cases
- When a package/version isn't found at all (#10623)
- When a version is found but its not the one the user expected (#7929)

The work we have planned for #10623 would make it so we have more information available to show for #7929.
